pomodoro test suite is too strict when checking for break-length, session-length and reset content

@hbar1st commented on [Tue Aug 21 2018](https://github.com/freeCodeCamp/freeCodeCamp/issues/18036)

<!-- NOTE: If you're reporting a security issue, don't create a GitHub issue. Instead, email security@freecodecamp.org. We will look into it immediately. -->

#### Describe your problem and - if possible - how to reproduce it
in this sample codepen created for the pomodoro clock challenge, the test suite fails to identify input fields that are being used to hold the break-length, session-length and to house the reset button.
https://codepen.io/hbar1st/pen/XPJKKo?editors=1100

This tells me the suite checks are too narrow in scope as they do not allow the use of form input fields which seem to be a natural fit for the pomodoro clock requirements.

The error for break-length for eg. states it cannot find a value of 5 even though *literally* the value attribute of the input tag is set:
Here's the start of one of the errors:
I can see an element, with corresponding id="break-length", which by default (on load) displays a value of 5. A value of 5 is not displayed by default: expected '' to equal '5' 
AssertionError: A value of 5 is not displayed by default: expected '' to equal '5'
    at o.c (https://cdn.freecodecamp.org/testable-projects-fcc/v1/bundle.js:205:1134)

#### Add a Link to the page with the problem
https://codepen.io/hbar1st/pen/XPJKKo?editors=1100

#### Tell us about your browser and operating system

* Browser Name: Chrome
* Browser Version: Version 68.0.3440.106
* Operating System: Windows 10


#### If possible, add a screenshot here

--- a/src/project-tests/calculator-tests.js
+++ b/src/project-tests/calculator-tests.js
@@ -1,5 +1,5 @@
 import { assert } from 'chai';
-import { clickButtonsById } from '../utils/element-utils';
+import { clickButtonsById, getInputValue } from '../utils/element-utils';
 import { frontEndLibrariesStack } from '../utils/shared-test-strings';
 
 export default function createCalculatorTests() {
@@ -55,8 +55,7 @@ export default function createCalculatorTests() {
         assert.isNotNull(element);
 
         assert.strictEqual(
-          element.nodeName.toLowerCase() === 'input' ?
-            element.value : element.innerHTML,
+          getInputValue(element),
           '=',
           'The element with id="equal" should have an "=" as its content '
         );
@@ -146,8 +145,7 @@ export default function createCalculatorTests() {
         );
 
         assert.strictEqual(
-          element.nodeName.toLowerCase() === 'input' ?
-            element.value : element.innerHTML,
+          getInputValue(element),
           '.',
           'The element with id="decimal" should have "." as its content '
         );
@@ -180,7 +178,7 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_5, _x, _1, _plus, _5, _plus, _9, _2, _eq, _AC]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '0',
           'Element with with id="display" should show 0 '
         );
@@ -192,7 +190,7 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_1, _2, _3]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '123',
           'Numbers do not display correctly within id="display" '
         );
@@ -205,8 +203,8 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_3, _plus, _5, _x, _6, _min, _2, _div, _4, _eq]);
         assert(
-          document.getElementById('display').innerHTML === '32.5' ||
-          document.getElementById('display').innerHTML === '11.5',
+          getInputValue(document.getElementById('display')) === '32.5' ||
+          getInputValue(document.getElementById('display')) === '11.5',
           `The expression 3 + 5 * 6 - 2 / 4 should produce 32.5 or 11.5 as an
           answer, depending on the logic your calculator uses
           (formula vs. immediate execution) `
@@ -214,7 +212,7 @@ export default function createCalculatorTests() {
         clearDisplay();
         clickButtonsById([_5, _min, _9, _plus, _5, _eq]);
         assert(
-          document.getElementById('display').innerHTML === '1',
+          getInputValue(document.getElementById('display')) === '1',
           'The expression 5 - 9 + 5 should produce a result of 1 '
         );
       });
@@ -225,7 +223,7 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_0, _0, _0]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '0',
           'An input of 0 0 0 should display 0 '
         );
@@ -238,14 +236,14 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_5, _dec, _dec, _0]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '5.0',
           'An input of "5 . . 0" should display 5.0 '
         );
         clearDisplay();
         clickButtonsById([_5, _dec, _5, _dec, _5]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '5.55',
           'An input of "5 . 5 . 5" should display 5.55 '
         );
@@ -257,28 +255,28 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_1, _0, _dec, _5, _min, _5, _dec, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '5',
           'The expression "10.5 - 5.5" should produce an output of "5" '
         );
         clearDisplay();
         clickButtonsById([_5, _x, _5, _dec, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '27.5',
           'The expression "5 * 5.5" should produce an output of "27.5" '
         );
         clearDisplay();
         clickButtonsById([_1, _0, _dec, _5, _plus, _5, _dec, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '16',
           'The expression "10.5 + 5.5" should produce an output of "16" '
         );
         clearDisplay();
         clickButtonsById([_1, _0, _div, _2, _dec, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '4',
           'The expression "10 / 2.5" should produce an output of "4" '
         );
@@ -290,14 +288,14 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_5, _x, _min, _plus, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '10',
           'The sequence "5 * - + 5" = should produce an output of "10" '
         );
         clearDisplay();
         clickButtonsById([_5, _plus, _plus, _5, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '10',
           'The sequence "5 + + 5" = should produce an output of "10" '
         );
@@ -310,14 +308,14 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_5, _min, _2, _eq, _div, _2, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '1.5',
           'The sequence "5 - 2 = / 2 =" should produce an output of "1.5" '
         );
         clearDisplay();
         clickButtonsById([_5, _plus, _5, _eq, _plus, _3, _eq]);
         assert.strictEqual(
-          document.getElementById('display').innerHTML,
+          getInputValue(document.getElementById('display')),
           '13',
           'The sequence "5 + 3 = + 3 =" should produce an output of "13" '
         );
@@ -331,7 +329,7 @@ export default function createCalculatorTests() {
       function() {
         clickButtonsById([_2, _div, _7, _eq]);
         assert.isOk(/0?\.2857\d*/.test(
-          document.getElementById('display').innerHTML),
+          getInputValue(document.getElementById('display'))),
           'The expression "2 / 7" should produce an output number with at ' +
           'least 4 decimal places of precision '
         );

--- a/src/project-tests/pomodoro-clock-tests.js
+++ b/src/project-tests/pomodoro-clock-tests.js
@@ -1,5 +1,5 @@
 import { assert } from 'chai';
-import { clickButtonsById } from '../utils/element-utils';
+import { clickButtonsById, getInputValue } from '../utils/element-utils';
 import { frontEndLibrariesStack } from '../utils/shared-test-strings';
 
 export default function createPomodoroClockTests() {
@@ -196,7 +196,7 @@ export default function createPomodoroClockTests() {
       function() {
         const breakLength = document.getElementById('break-length');
         assert.strictEqual(
-          breakLength.innerHTML,
+          getInputValue(breakLength),
           '5',
           'A value of 5 is not displayed by default'
         );
@@ -208,7 +208,7 @@ export default function createPomodoroClockTests() {
       function() {
         const sessionLength = document.getElementById('session-length');
         assert.strictEqual(
-          sessionLength.innerHTML,
+          getInputValue(sessionLength),
           '25',
           'A value of 25 is not displayed by default'
         );
@@ -291,10 +291,14 @@ export default function createPomodoroClockTests() {
                 // (pomodoro is stopped)
                 setTimeout(() => {
                   const breakLenAfterResetCorrect = (
-                    document.getElementById('break-length').innerHTML === '5'
+                    getInputValue(
+                      document.getElementById('break-length')
+                    ) === '5'
                   );
                   const sessionLenAfterResetCorrect = (
-                    document.getElementById('session-length').innerHTML === '25'
+                    getInputValue(
+                      document.getElementById('session-length')
+                    ) === '25'
                   );
                   if (
                     !breakLenAfterResetCorrect || !sessionLenAfterResetCorrect
@@ -335,13 +339,13 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById([breakMin, breakMin, breakMin, breakMin]);
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML,
+          getInputValue(document.getElementById('break-length')),
           '1'
         );
         resetTimer();
         clickButtonsById([breakMin]);
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML,
+          getInputValue(document.getElementById('break-length')),
           '4'
         );
       });
@@ -353,13 +357,13 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById(Array(4).fill(breakPlus));
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML,
+          getInputValue(document.getElementById('break-length')),
           '9'
         );
         resetTimer();
         clickButtonsById([breakPlus]);
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML, '6'
+          getInputValue(document.getElementById('break-length')), '6'
         );
       });
 
@@ -370,13 +374,13 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById(Array(4).fill(seshMin));
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '21'
         );
         resetTimer();
         clickButtonsById([seshMin]);
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '24'
         );
       });
@@ -388,13 +392,13 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById(Array(4).fill(seshPlus));
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '29'
         );
         resetTimer();
         clickButtonsById([seshPlus]);
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '26'
         );
       });
@@ -405,14 +409,14 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById(Array(10).fill(breakMin));
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML,
+          getInputValue(document.getElementById('break-length')),
           '1',
           'Value in element with id of "break-length" is less than 1.'
         );
         resetTimer();
         clickButtonsById(Array(30).fill(seshMin));
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '1',
           'Value in element with id of "session-length" is less than 1.'
         );
@@ -424,14 +428,14 @@ export default function createPomodoroClockTests() {
       function() {
         clickButtonsById(Array(60).fill(breakPlus));
         assert.strictEqual(
-          document.getElementById('break-length').innerHTML,
+          getInputValue(document.getElementById('break-length')),
           '60',
           'Value in element with id of "break-length" is greater than 60.'
         );
         resetTimer();
         clickButtonsById(Array(40).fill(seshPlus));
         assert.strictEqual(
-          document.getElementById('session-length').innerHTML,
+          getInputValue(document.getElementById('session-length')),
           '60',
           'Value in element with id of "session-length" is greater than 60.'
         );
@@ -448,7 +452,7 @@ export default function createPomodoroClockTests() {
           getMinutes(
             document.getElementById('time-left').innerHTML
           ),
-          document.getElementById('session-length').innerHTML
+          getInputValue(document.getElementById('session-length'))
         );
       });
 
@@ -611,7 +615,7 @@ export default function createPomodoroClockTests() {
                   'Test timed out because Break time didn\'t start with the ' +
                   'correct value: ' + (
                     parseInt(getMinutes(timeLeft.innerHTML), 10) + 1
-                  ) + ' instead of ' + breakLength.innerHTML
+                  ) + ' instead of ' + getInputValue(breakLength)
                 )
               );
               observer.disconnect();
@@ -660,7 +664,7 @@ export default function createPomodoroClockTests() {
                   'Timer has switched to Break time, but it didn\'t start ' +
                   'with the correct value: ' + (
                     parseInt(getMinutes(timeLeft.innerHTML), 10) + 1
-                  ) + ' instead of ' + breakLength.innerHTML
+                  ) + ' instead of ' + getInputValue(breakLength)
                 )
               );
               observer.disconnect();
@@ -668,7 +672,7 @@ export default function createPomodoroClockTests() {
             } else if (shouldBeInBreak) {
               if (currentTimer.innerHTML !== sessionLabel) {
                 let getTimeLeftHTML = +getMinutes(timeLeft.innerHTML);
-                if (getTimeLeftHTML === +breakLength.innerHTML) {
+                if (getTimeLeftHTML === +getInputValue(breakLength)) {
                   resolve();
                 } else {
                   reject(
@@ -676,7 +680,7 @@ export default function createPomodoroClockTests() {
                       'Timer has switched to Break time, but it didn\'t ' +
                       'start with the correct value: ' +
                       getMinutes(timeLeft.innerHTML) + ' instead of ' +
-                      breakLength.innerHTML
+                      getInputValue(breakLength)
                     )
                   );
                   observer.disconnect();
@@ -771,7 +775,7 @@ export default function createPomodoroClockTests() {
               const sessionLength = document.getElementById('session-length');
               if (currentTimer.innerHTML !== breakLabel) {
                 let getTimeLeftHTML = +getMinutes(timeLeft.innerHTML);
-                if (getTimeLeftHTML === +sessionLength.innerHTML) {
+                if (getTimeLeftHTML === +getInputValue(sessionLength)) {
                   resolve();
                 } else {
                   reject(
@@ -779,7 +783,7 @@ export default function createPomodoroClockTests() {
                       'Timer has switched back to Session time, but it ' +
                       'didn\'t start with the correct value: ' +
                       getMinutes(timeLeft.innerHTML) + ' instead of ' +
-                      sessionLength.innerHTML
+                      getInputValue(sessionLength)
                     )
                   );
                 }
@@ -821,7 +825,7 @@ export default function createPomodoroClockTests() {
                   'Test timed out because Break time didn\'t start with ' +
                   'the correct value: ' + (
                     parseInt(getMinutes(timeLeft.innerHTML), 10) + 1
-                  ) + ' instead of ' + breakLength.innerHTML
+                  ) + ' instead of ' + getInputValue(breakLength)
                 )
               );
             } else if (timeLeft.innerHTML === '00:00') {

--- a/src/utils/element-utils.js
+++ b/src/utils/element-utils.js
@@ -45,3 +45,8 @@ export function hasUniqueColorsCount(elements, numberOfColors) {
 
   return false;
 }
+
+export function getInputValue(element) {
+  return element.nodeName.toLowerCase() === 'input' ?
+    element.value : element.innerText;
+}

